libraries/Microsoft.Bot.Builder/TokenResolver.cs

```diff
@@ -0,0 +1,134 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Security.Claims;
+using System.Security.Principal;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder
+{
+    internal class TokenResolver
+    {
+        private const int PollingInterval = 1000;   // Poll for token every 1 second.
+
+        public static void CheckForOAuthCards(BotFrameworkAdapter adapter, ITurnContext turnContext, Activity activity, CancellationToken cancellationToken)
+        {
+            if (activity?.Attachments != null)
+            {
+                foreach (var attachment in activity?.Attachments?.Where(a => a.ContentType == OAuthCard.ContentType))
+                {
+                    OAuthCard oauthCard = attachment.Content as OAuthCard;
+                    if (oauthCard != null && !string.IsNullOrEmpty(oauthCard.ConnectionName))
+                    {
+                        // Poll as a background task
+                        Task.Run(() => PollForTokenAsync(adapter, turnContext, activity, oauthCard.ConnectionName, cancellationToken));
+                    }
+                }
+            }
+        }
+
+        private static async Task PollForTokenAsync(BotFrameworkAdapter adapter, ITurnContext turnContext, Activity activity, string connectionName, CancellationToken cancellationToken)
+        {
+            TokenResponse tokenResponse = null;
+            bool shouldEndPolling = false;
+            int pollingTimeout = TurnStateConstants.OAuthLoginTimeoutMsValue;
+            int pollingRequestsInterval = PollingInterval;
+            var loginTimeout = turnContext?.TurnState?.Get<object>(TurnStateConstants.OAuthLoginTimeoutKey) as int?;
+
+            // Override login timeout with value set from the OAuthPrompt or by the developer
+            if (loginTimeout != null)
+            {
+                pollingTimeout = (int)loginTimeout;
+            }
+
+            if (activity.From == null || string.IsNullOrWhiteSpace(activity.From.Id))
+            {
+                throw new ArgumentNullException("TokenResolver.PollForTokenAsync: missing Activity from or from.id");
+            }
+
+            if (string.IsNullOrWhiteSpace(connectionName))
+            {
+                throw new ArgumentNullException(nameof(connectionName));
+            }
+
+            var stopwatch = Stopwatch.StartNew();
+
+            while (stopwatch.Elapsed < TimeSpan.FromMilliseconds(pollingTimeout) && !shouldEndPolling)
+            {
+                tokenResponse = await adapter.GetUserTokenAsync(turnContext, connectionName, null, cancellationToken).ConfigureAwait(false);
+
+                if (tokenResponse != null)
```

Comment:
The null checking through here makes this code really complex. IS there a way to re-write this w/o all of the null programming?

Reply:
This method on the adapter can return null so we have to deal with that case. It means "the token isn't there yet" and is normal.